phpstan.el

```diff
@@ -28,19 +28,233 @@
+it returns the value of `SOURCE' as it is."
+  (let ((root-directory (expand-file-name (php-project-get-root-dir)))
+        (prefix
+         (cond
+          ((not (null phpstan-replace-path-prefix)) phpstan-replace-path-prefix)
+          ((eq 'docker phpstan-executable) "/app")
+          ((and (consp phpstan-executable)
+                (string= "docker" (car phpstan-executable))) "/app"))))
+    (if prefix
+        (expand-file-name
+         (replace-regexp-in-string (concat "\\`" (regexp-quote root-directory))
+                                   ""
+                                   source-original t t)
+         prefix)
+      (or source source-original))))
+
+(defun phpstan-get-level ()
+  "Return path to phpstan configure file or `NIL'."
+  (cond
+   ((null phpstan-level) "0")
+   ((integerp phpstan-level) (int-to-string phpstan-level))
+   ((symbolp phpstan-level) (symbol-name phpstan-level))
+   (t phpstan-level)))
+
+(defun phpstan-get-executable ()
+  "Return PHPStan excutable file and arguments."
+  (cond
+   ((eq 'docker phpstan-executable)
+    (list "run" "--rm" "-v"
+          (concat (expand-file-name (php-project-get-root-dir)) ":/app")
+          "phpstan/phpstan"))
+   ((and (consp phpstan-executable)
+         (eq 'root (car phpstan-executable)))
+    (expand-file-name (cdr phpstan-executable) (php-project-get-root-dir)))
+   ((and phpstan-flycheck-auto-set-executable
+         (listp phpstan-executable)
+         (stringp (car phpstan-executable))
+         (listp (cdr phpstan-executable)))
+    (cdr phpstan-executable))
+   ((null phpstan-executable)
+    (let ((vendor-phpstan (expand-file-name "vendor/bin/phpstan"
+                                            (php-project-get-root-dir))))
+      (cond
+       ((file-exists-p vendor-phpstan) (list vendor-phpstan))
+       ((executable-find "phpstan") (list (executable-find "phpstan")))
+       (t (error "PHPStan executable not found")))))))
+
+(defun phpstan-get-command-args ()
+  "Return command line argument for PHPStan."
+  (let ((executable (phpstan-get-executable))
+        (args (list "analyze" "--errorFormat=raw" "--no-progress" "--no-interaction"))
+        (path (phpstan-normalize-path (phpstan-get-config-file)))
+        (level (phpstan-get-level)))
+    (when path
+      (setq args (append args (list "-c" path))))
+    (when level
+      (setq args (append args (list "-l" level))))
+    (append executable args)))
+
 ;;;###autoload
 (when (featurep 'flycheck)
-  (flycheck-define-checker phpstan-checker
+  (flycheck-define-checker phpstan
     "PHP static analyzer based on PHPStan."
-    :command ("phpstan" 
-              "analyze" 
-              "--no-progress"
-              "--errorFormat=raw" 
-              source)
-    :working-directory (lambda (_) (php-project-get-root-dir))
-    :enabled (lambda () (locate-dominating-file "phpstan.neon" default-directory))
+    :command ("php" (eval (phpstan-get-command-args))
```

Comment:
@Fuco1  This is an insane hack.

[`phpstan-enabled-and-set-flycheck-variable`](https://github.com/emacs-php/phpstan.el/blob/b0665de9a05df4ea60ca713a990bbd534124dd3f/phpstan.el#L169-L183) invoked as `:enabled` function sets `flycheck-phpstan-executable` as its side effect when `flycheck-phpstan-executable` is not set.  Since it uses the fact that `:enabled` is called beforehand, there is a possibility that it will be destroyed in the future with low probability.

It was possible to explicitly write `flycheck-phpstan-executable` in the user setting, but it was handled within this function as it is somewhat complicated.  I am looking for ways to solve this problem besides this hack.